bson/src/main/org/bson/codecs/pojo/LazyPropertyModelCodec.java

```diff
@@ -163,19 +163,41 @@ private <V> PropertyModel<V> getSpecializedPropertyModel(final PropertyModel<V>
     static final class NeedSpecializationCodec<T> extends PojoCodec<T> {
         private final ClassModel<T> classModel;
         private final DiscriminatorLookup discriminatorLookup;
+        private final CodecRegistry codecRegistry;
 
-        NeedSpecializationCodec(final ClassModel<T> classModel, final DiscriminatorLookup discriminatorLookup) {
+        NeedSpecializationCodec(final ClassModel<T> classModel, final DiscriminatorLookup discriminatorLookup, final CodecRegistry codecRegistry) {
             this.classModel = classModel;
             this.discriminatorLookup = discriminatorLookup;
+            this.codecRegistry = codecRegistry;
         }
 
         @Override
-        public T decode(final BsonReader reader, final DecoderContext decoderContext) {
-            throw exception();
+        public void encode(final BsonWriter writer, final T value, final EncoderContext encoderContext) {
+            tryEncode(codecRegistry.get(value.getClass()), writer, value, encoderContext);
         }
 
         @Override
-        public void encode(final BsonWriter writer, final T value, final EncoderContext encoderContext) {
+        public T decode(final BsonReader reader, final DecoderContext decoderContext) {
+            return tryDecode(reader, decoderContext);
+        }
+
+        @SuppressWarnings("unchecked")
+        private <A> void tryEncode(final Codec<A> codec,  final BsonWriter writer, final T value, final EncoderContext encoderContext) {
+            try {
```

Comment:
Looks like `codec.encode` can be cyclical if `codecRegistry.get(value.getClass())` returns a `LazyCodec` - eg there is no codec available for that class at the time.

This leads to a stackoverflow of:

```
(LazyCodec) codecRegistry.get(value.getClass())
LazyCodec.encode(...) // Creates a new LazyPropertyModelCodec and round we go again.
```

This can be fixed by checking here if the class of the value and the type of the classModel is the same - if it is it needs specialization, so throw.